README.md
Benchwarmer
===========

Benchwarmer is an Elixir micro-benchmarking utility that runs a function (or list of functions) repeatedly against a dataset for a period of time, and then reports on the average time each operation took to complete, allowing for easy comparison.

Benchwarmer was inspired by the built-in benchmark operations in the Go test library.


Installation
------------
Add Benchwarmer to your `mix.exs` dependencies with whatever published semantic version you'd like to depend on.

```elixir
def deps do
  [ { :benchwarmer, "~> x.y.z" } ]
end
```


Usage
-----
You can simply pass an inline function, results will be pretty printed to screen as well as returned in a struct:

```iex
iex> Benchwarmer.benchmark fn -> 123456*654321 end
*** #Function<20.90072148/0 in :erl_eval.expr/5> ***
1.2 sec     2M iterations   0.61 μs/op

[%Benchwarmer.Results{...}]
```

Comparing two different functions with the same data as an argument:

```iex
iex> alphabet = "abcdefghijklmnopqrstuvwxyz"
iex> Benchwarmer.benchmark [&String.first/1, &String.last/1], alphabet
*** &String.first/1 ***
1.9 sec     8M iterations   0.24 μs/op

*** &String.last/1 ***
1.9 sec   524K iterations   3.75 μs/op

[%Benchwarmer.Results{...}, %Benchwarmer.Results{...}]
```


Compared to other Elixir benchmarking tools
-------------------------------------------
For TDD-style benchmarking, you should probably check out [Benchfella][bf], which I discovered after starting this project.  It's quite good.

This library is a little more focused on quick comparative benchmarks in IEx, working without adding a single line of code to your actual project files.


Work in Progress
----------------
Please note this library is a work in progress, and the API will likely change.

Pull requests are welcome!

[bf]: https://github.com/alco/benchfella


CHANGELOG.md
# Changelog

## 0.0.2 (8 September 2014)

 * Update dependencies for for Elixir 1.0 release.

## 0.0.1 (25 July 2014)

 * Initial release available in hex.


test/benchwarmer_test.exs
defmodule BenchwarmerTest do
  use ExUnit.Case

  test "the truth" do
    assert 1 + 1 == 2
  end
end


mix.exs
defmodule Benchwarmer.Mixfile do
  use Mix.Project

  def project do
    [
      app:          :benchwarmer,
      version:      "0.0.2",
      elixir:       "~> 1.0.0-rc1",
      deps:         deps,
      name:         "Benchwarmer",
      source_url:   "https://github.com/mroth/benchwarmer",
      description:  description,
      package:      package
    ]
  end

  defp description do
    """
    Benchwarmer is an Elixir micro-benchmarking utility that runs a function (or
    list of functions) repeatedly against a dataset for a period of time, and
    then reports on the average time each operation took to complete, allowing
    for easy comparison.
    """
  end

  defp package do
    [
      contributors: [ "Matthew Rothenberg <mroth@mroth.info>" ],
      licenses:     [ "WTFPL" ],
      links:        %{
                       "GitHub" => "https://github.com/mroth/benchwarmer",
                       "Docs"   => "https://mroth.github.io/benchwarmer/"
                    }
    ]
  end

  # Configuration for the OTP application
  #
  # Type `mix help compile.app` for more information
  def application do
    [applications: []]
  end

  # Dependencies can be hex.pm packages:
  #
  #   {:mydep, "~> 0.3.0"}
  #
  # Or git/path repositories:
  #
  #   {:mydep, git: "https://github.com/elixir-lang/mydep.git", tag: "0.1"}
  #
  # Type `mix help deps` for more examples and options
  defp deps do
    [
      {:ex_doc,  github: "elixir-lang/ex_doc", only: :dev},
      {:earmark, github: "pragdave/earmark",   only: :dev}
    ]
  end
end


lib/benchwarmer/results.ex
defmodule Benchwarmer.Results do
  @moduledoc """
  Defines results of a Benchwarmer benchmark run.

  Handled as a struct with the following values:

    - n:             completed iterations
    - prev_n:        iterations run in previous cycle
    - duration:      time elapsed in μs
    - function:      the function that was executed when the benchmark was run
    - args:          the args that were used to execute the function
  """
  defstruct(
    n: 0, prev_n: 0, duration: 0, function: nil, args: nil
  )

  alias Benchwarmer.Results

  @doc """
  Returns average execution time per operation for a result, in microseconds.
  """
  @spec optime(Results) :: float
  def optime(%Results{n: 0}), do: 0.0
  def optime(%Results{duration: duration, n: n}), do: duration / n

end


# ########################################################################
# # On second thought, I think that overriding inspect makes it harder for
# # people who would be using this programmatically.
# ########################################################################
# defimpl Inspect, for: Benchwarmer.Results do
#   alias Benchwarmer.Results.Helpers
#
#   def inspect(r, _) do
#     name = "#{Inspect.inspect(r.function, nil)}"
#     content = Enum.join([
#                           "#{Helpers.pretty_time(r.duration)} sec",
#                           "#{Helpers.pretty_num(r.n)} iterations",
#                           "#{Helpers.pretty_optime(r)} μs/op"
#                         ], ", ")
#
#     """
#     #Benchwarmer.Results[#{name}]<#{content}>
#     """ |> String.rstrip
#   end
# end

defimpl String.Chars, for: Benchwarmer.Results do
  alias Benchwarmer.Results.Helpers

  def to_string(r) do
    name = "#{Inspect.inspect(r.function, nil)}"
    data = Enum.join [ "#{Helpers.pretty_time(r.duration)} sec",
                       "#{Helpers.pretty_num(r.n)} iterations",
                       "#{Helpers.pretty_optime(r)} μs/op" ], "   "

    """
    *** #{name} ***
    #{data}
    """
  end
end


defmodule Benchwarmer.Results.Helpers do
  @moduledoc false
  # Helper methods used to format Benchwarmer.Results when rendered to string or
  # inspected in iex.
  #
  # Mostly one-off string formatters. You can safely ignore these. :-)

  def pretty_optime(r) do
    Float.ceil(Benchwarmer.Results.optime(r), 2)
  end

  def pretty_time(ms) do
    Float.floor(ms/1_000_000, 1) |> Kernel.to_string
  end

  def pretty_num(n) do
    cond do
      n >= 1_000_000_000_000 ->
        Kernel.to_string(trunc(Float.floor(n/1_000_000_000_000))) <> "T"
      n >= 1_000_000_000 ->
        Kernel.to_string(trunc(Float.floor(n/1_000_000_000))) <> "B"
      n >= 1_000_000 ->
        Kernel.to_string(trunc(Float.floor(n/1_000_000))) <> "M"
      n >= 1000 ->
        Kernel.to_string(trunc(Float.floor(n/1000))) <> "K"
      true -> Kernel.to_string(n)
    end
    |> String.rjust(4)
  end

end


lib/benchwarmer.ex
defmodule Benchwarmer do
  @moduledoc """
  Benchwarmer is an Elixir micro-benchmarking utility that runs a function (or
  list of functions) repeatedly against a dataset for a period of time, and then
  reports on the average time each operation took to complete, allowing for easy
  comparison.

  It is optimized for interactive testing in IEx.

  Benchwarmer was inspired by the built-in benchmark operations in the Go
  test library.
  """

  @default_duration 1_000_000 # 1 second (1Mμs)

  alias Benchwarmer.Results


  @doc """
  Benchmarks a function or list of functions with optional args.

  ## Examples
  You can simply pass an inline function, results will be pretty printed to
  screen as well as returned in a struct:

      iex> Benchwarmer.benchmark fn -> 123456*654321 end
      #Function<20.90072148/0 in :erl_eval.expr/5>
      1.2 sec     2M iterations   0.61 μs/op

      [%Benchwarmer.Results{...}]

  Comparing two different functions with the same data as an argument:

      iex> alphabet = "abcdefghijklmnopqrstuvwxyz"
      iex> Benchwarmer.benchmark [&String.first/1, &String.last/1], alphabet
      *** &String.first/1 ****
      1.9 sec     8M iterations   0.24 μs/op

      *** &String.last/1 ****
      1.9 sec   524K iterations   3.75 μs/op

      [%Benchwarmer.Results{...}, %Benchwarmer.Results{...}]
  """
  def benchmark(f, args \\ [], min_duration \\ @default_duration) do
    functions = List.wrap(f)
    safe_args = List.wrap(args)

    Enum.map(functions, fn(fp) ->
      results = do_benchmark(fp, safe_args, min_duration)
      IO.puts results
      results
    end)

    # for future reference, gvaughn one-liner without the IO
    # List.wrap(f) |> Enum.map &do_benchmark(&1, List.wrap(args), min_duration)
  end

  #
  # Benchmarks an individual function against args and returns results.
  #
  defp do_benchmark(f, args, min_duration, results \\ %Results{}) do
    # TODO: does this help with the load timer issues? check it out.
    :timer.start()

    cond do
      # when elapsed test time is greater than minimum, return results
      # (decorate results with original function for later introspection)
      results.duration >= min_duration -> %{results | function: f, args: args}

      # first pass, run for a single iteration only
      results.n == 0 ->
        {:ok, _, optime} = run_n_times(f, args, 1)
        do_benchmark( f, args, min_duration,
                      %{results | n: 1, prev_n: 1, duration: optime}
                    )

      # default case, run with double iterations of previous instance
      true ->
        {:ok, iters, optime} = run_n_times(f, args, results.prev_n * 2)
        do_benchmark( f, args, min_duration,
                      %{results | n: results.n + iters,
                                  prev_n: iters,
                                  duration: results.duration + optime
                      }
                    )
    end
  end

  #
  # Runs a function with args N times.
  #
  defp run_n_times(f, args, 1) do
    {optime, _results} = :timer.tc(f, args)
    {:ok, 1, optime}
  end
  defp run_n_times(f, args, n) do
    start = :erlang.now
    for _n <- 1..n, do: apply(f, args)
    stop = :erlang.now
    optime = :timer.now_diff(stop, start)
    {:ok, n, optime}
  end

end
